cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -472,7 +472,7 @@ func GetAvailableResources(clientBuilder clientbuilder.ControllerClientBuilder)
 	discoveryClient := client.Discovery()
 	_, resourceMap, err := discoveryClient.ServerGroupsAndResources()
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("unable to get all supported resources from server: %v", err))
+		return nil, fmt.Errorf("unable to get all supported resources from server: %v", err)
```

Comment:
this was this way initially, and changed here https://github.com/kubernetes/kubernetes/pull/49495 by @deads2k 

/assign @deads2k

Reply:
yeah, turning this back into a fatal error means a cluster with unhealthy APIServices that require the controller manager to start up to spin up backing pods is in a deadlock

Reply:
Hmm, make sense.

The deadlock only happens when controller-manager and apiservices pods are down at the same time, right? This function `GetAvailableResources()` is only called once during kube-controller-manager startup.

I'm wondering whether there is a way to fix both worlds. As is, things like https://github.com/kubernetes/kubernetes/issues/107583 might happen without even being noticed. And I also see a similar issue in our setup where a controller wasn't started at all while the main controller-manager still runs fine.

Basically, there are multiple reasons this call `discoveryClient.ServerGroupsAndResources()` might fail:

1. A temporary network hiccup
2. Server side ratelimiting. (with APF enabled, and discovery API calls usually come in a burst)

In any case, if the results can't be trusted, it is hard to write check correctly. For e.g, clusterroleaggregation controller / hpa controller may not start at all.

```go
func startClusterRoleAggregrationController(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {
	if !controllerContext.AvailableResources[schema.GroupVersionResource{Group: "rbac.authorization.k8s.io", Version: "v1", Resource: "clusterroles"}] {
		return nil, false, nil
	}
	...
}

func startHPAController(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {
	if !controllerContext.AvailableResources[schema.GroupVersionResource{Group: "autoscaling", Version: "v1", Resource: "horizontalpodautoscalers"}] {
		return nil, false, nil
	}
	...
}
```

I wonder whether it is reasonable to treat APIServices unavailable as a special case (ignoring any 503s) in `discoveryClient.ServerGroupsAndResources()` while any other errors should be considered as fatal.